[CCIP-2958] Add price reader config and bindings for the TokenPriceChainSelector chain

asoliman92 opened the pull request:
Add price reader config and bindings for the TokenPriceChainSelector chain

asoliman92 committed:
Add price reader config and bindings for the TokenPriceChainSelector chain

asoliman92 committed:
fix linting

asoliman92 committed:
Add mock aggregator to home chain and add offchain config to commit plugin in integration test

asoliman92 committed:
Separate price reader config

* Added a merge function to merge config readers.
* If the chain is the chain that has price aggregator then we merge it into the chain (whether it's a dest or src chain)

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

asoliman92 committed:
Add decimals to commit offchain config

asoliman92 committed:
go mod tidy

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

makramkd commented on core/capabilities/ccip/ccip_integration_tests/helpers.go:
```diff
@@ -267,6 +270,8 @@ func createUniverses(
 
 		universes[chainID] = universe
 	}
+	// Set up home chain first
+	homeChainUniverse := setupHomeChain(t, homeChainBase.owner, homeChainBase.backend)
```
Why change the call site / ordering?

asoliman92 replied:
No need 👍. I thought initially we might need the token contracts before doing extra stuff but that's not the case.

makramkd commented on core/capabilities/ccip/ccip_integration_tests/helpers.go:
```diff
@@ -386,12 +391,18 @@ func setupHomeChain(t *testing.T, owner *bind.TransactOpts, backend *backends.Si
 	require.NoError(t, err, "failed to add node operator to the capability registry")
 	backend.Commit()
 
+	aggAddr, _, aggContract, err := mock_v3_aggregator_contract.DeployMockV3AggregatorContract(owner, backend, 18, big.NewInt(5e18))
```
Maybe make the decimals (18) a constant so that you can reference the right value here and in the offchain config

makramkd commented on core/capabilities/ccip/ccip_integration_tests/ocr_node_helper.go:
```diff
@@ -267,17 +267,26 @@ func createConfigV2Chain(chainID *big.Int) *v2toml.EVMConfig {
 			case <-tick.C:
 				for _, uni := range universes {
+					if uni.chainID == homeChainID {
+						agg := homeChainUni.mockAggregator
+						_, err := agg.UpdateRoundData(uni.owner, big.NewInt(round), big.NewInt(5e18), big.NewInt(time.Now().UTC().UnixNano()), big.NewInt(time.Now().UTC().UnixNano()))
+						if err != nil {
+							panic(fmt.Errorf("unable to update round data: %w", err))
+						}
```
Why update to the same value?

asoliman92 replied:
Added randomization factor. I don't think it's very helpful though as there's no direct way to test it in the integration test. Do you have a specific assertion to do for it in the test?

makramkd replied:
At least for this change being merged in I'm not sure this is necessary since the deviation checking is not being done yet in the plugin. So I feel like this can come later.

makramkd commented on core/capabilities/ccip/launcher/launcher_test.go:
```diff
@@ -489,6 +522,24 @@ func Test_launcher_processDiff(t *testing.T) {
+			RemoteGasPriceBatchWriteFrequency: *commonconfig.MustNewDuration(time.Second),
+			TokenPriceBatchWriteFrequency:     *commonconfig.MustNewDuration(time.Second),
+			PriceSources: map[ocrtypes.Account]pluginconfig.ArbitrumPriceSource{
+				"0xtokenAddress": {
+					AggregatorAddress: "0xaggregatorAddress",
```
Hmm you should not be able to put any random string here, it should be a valid ethereum address. Looks like we missed calling `Validate` on the `ArbitrumPriceSource` in chainlink-ccip

asoliman92 replied:
Good catch! We were not validating configs when creating oracles. Adding them.

makramkd commented on core/capabilities/ccip/oraclecreator/inprocess.go:
```diff
@@ -159,7 +159,10 @@ func (i *inprocessOracleCreator) CreatePluginOracle(pluginType cctypes.PluginTyp
 	}
+
 	var execBatchGasLimit uint64
+	var commitOffchainConfig *pluginconfig.CommitOffchainConfig
```
Hmm this func is starting to get pretty hard to read. We should clean it up a bit, in a followup PR

asoliman92 replied:
Agreed. 
https://smartcontract-it.atlassian.net/browse/CCIP-3039

asoliman92 committed:
Fix tests after adding offchain validation

asoliman92 committed:
Fix linting

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

asoliman92 committed:
extract offchain config into a function in helpers

asoliman92 committed:
Fix ocr configs validation in launcher

ocr3_node_test passes but launcher tests failing because they're not using the correct offchain config

asoliman92 committed:
Move the ccip offchain validation to in porcess instead of the launcher

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

asoliman92 committed:
modtidy

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

makramkd commented on core/capabilities/ccip/ccip_integration_tests/helpers.go:
```diff
@@ -450,6 +464,45 @@ func AddChainConfig(
+	if pluginType == cctypes.PluginTypeCCIPCommit {
+		encodedOffchainConfig, err2 = pluginconfig.EncodeCommitOffchainConfig(pluginconfig.CommitOffchainConfig{
+			RemoteGasPriceBatchWriteFrequency: *commonconfig.MustNewDuration(RemoteGasPriceBatchWriteFrequency),
+			TokenPriceBatchWriteFrequency:     *commonconfig.MustNewDuration(time.Second),
```
Lets pull out the frequency to a constant as well so it can be discovered and changed easily

makramkd commented on core/capabilities/ccip/ccip_integration_tests/helpers.go:
```diff
@@ -450,6 +464,45 @@ func AddChainConfig(
+		encodedOffchainConfig, err2 = pluginconfig.EncodeCommitOffchainConfig(pluginconfig.CommitOffchainConfig{
+			RemoteGasPriceBatchWriteFrequency: *commonconfig.MustNewDuration(RemoteGasPriceBatchWriteFrequency),
+			TokenPriceBatchWriteFrequency:     *commonconfig.MustNewDuration(time.Second),
+			PriceSources: map[ocrtypes.Account]pluginconfig.ArbitrumPriceSource{
+				ocrtypes.Account(uni.weth.Address().Hex()): {
+					AggregatorAddress: h.mockAggregatorAddress.Hex(),
+					DeviationPPB:      ccipocr3.BigInt{Int: big.NewInt(1)},
+				},
+			},
+			TokenDecimals: map[ocrtypes.Account]uint8{
+				ocrtypes.Account(uni.weth.Address().Hex()): 18,
+			},
+			TokenPriceChainSelector: getSelector(h.chainID),
```
Please make these changes in the integration test in integration-tests as well

makramkd commented on core/capabilities/ccip/ccip_integration_tests/helpers.go:
```diff
@@ -472,26 +525,10 @@ func (h *homeChain) AddDON(
-			require.NoError(t, err2)
-		}
+		encodedOffchainConfig := getPluginConfig(t, pluginType, uni, *h)
+		// commit: len:337, cap:352
+		// exec: len:171, cap:176
```
Can delete these lines

makramkd commented on core/capabilities/ccip/ccip_integration_tests/ocr_node_helper.go:
```diff
@@ -267,17 +268,31 @@ func createConfigV2Chain(chainID *big.Int) *v2toml.EVMConfig {
+						randomNumber := rand.Intn(5)
+						// answer = previous round answer + (previous round answer * random percentage)
+						answer = new(big.Int).Add(answer, new(big.Int).Div(new(big.Int).Mul(answer, big.NewInt(int64(randomNumber))), big.NewInt(100)))
+						_, err := agg.UpdateRoundData(uni.owner, big.NewInt(round), big.NewInt(5e18), big.NewInt(time.Now().UTC().UnixNano()), big.NewInt(time.Now().UTC().UnixNano()))
```
`answer` doesn't seem to be used in this `UpdateRoundData` call

asoliman92 committed:
Fix comments

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader

asoliman92 committed:
WIP add price feed support to deployment integration tests

asoliman92 committed:
Merge branch 'ccip-develop' into price-reader
